File: core/SubsetsII/SubsetsII.py
from typing import List
from copy import deepcopy


def subsetsWithDup(nums: List[int]) -> List[List[int]]:
    nums, length = sorted(nums), len(nums)
    res = []

    def dfs(start, remain, is_repeat):
        if start == length:
            res.append(deepcopy(remain))
            return

        dfs(start + 1, remain, False)
        if not is_repeat and start > 0 and nums[start] == nums[start - 1]:
            return
        remain.append(nums[start])
        dfs(start + 1, remain, True)
        remain.pop()

    dfs(0, [], True)

    return res

File: core/SubsetsII/SubsetsII.java
package com.dbc;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SubsetsII {
    private final List<List<Integer>> res = new ArrayList<>();

    private void dfs(int start, int[] nums, List<Integer> remain, boolean select) {
        if (start == nums.length) {
            this.res.add(new ArrayList<>(remain));
            return;
        }
        dfs(start + 1, nums, remain, false);
        if (!select && start > 0 && nums[start] == nums[start - 1])
            return;
        remain.add(nums[start]);
        dfs(start + 1, nums, remain, true);
        remain.remove(remain.size() - 1);
    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        dfs(0, nums, new ArrayList<>(), false);
        return this.res;
    }
}
